src/main/java/org/manifold/compiler/front/Main.java

```diff
@@ -307,41 +272,49 @@ public Expression visitFunctionInvocationExpression(
+  public ExpressionVertex visitPrimitiveNodeDefinitionExpression(
       ManifoldParser.PrimitiveNodeDefinitionExpressionContext context) {
-    // extract port-mapping type
-    FunctionTypeValueContext typeCtx = context.functionTypeValue();
-    Expression typevalue = typeCtx.accept(this);
-    return new PrimitiveNodeDefinitionExpression(typevalue);
+    ExpressionVertex vSignature = context.functionTypeValue().accept(this);
+    ExpressionEdge eSignature = new ExpressionEdge(vSignature, null);
+    exprGraph.addEdge(eSignature);
+    PrimitiveNodeVertex vNode = new PrimitiveNodeVertex(exprGraph, eSignature);
+    exprGraph.addVertex(vNode);
+    return vNode;
   }
 
   @Override
-  public Expression visitPrimitivePortDefinitionExpression(
+  public ExpressionVertex visitPrimitivePortDefinitionExpression(
       ManifoldParser.PrimitivePortDefinitionExpressionContext context) {
-    // extract signal type
-    TypevalueContext typeCtx = context.typevalue();
-    Expression typevalue = typeCtx.accept(this);
-    // are there any attributes?
+
+    ExpressionVertex vSignalType = context.typevalue().accept(this);
+    ExpressionEdge eSignalType = new ExpressionEdge(vSignalType, null);
+    exprGraph.addEdge(eSignalType);
+
+    ExpressionVertex vAttributes;
     if (context.tupleTypeValue() != null) {
-      TupleTypeValueContext attributeTypesContext = context.tupleTypeValue();
-      Expression attributes = attributeTypesContext.accept(this);
-      return new PrimitivePortDefinitionExpression(typevalue, attributes);
+      vAttributes = context.tupleTypeValue().accept(this);
     } else {
-      return new PrimitivePortDefinitionExpression(typevalue);
+      vAttributes = new ConstantValueVertex(exprGraph, 
+          NilTypeValue.getInstance());
+      exprGraph.addVertex(vAttributes);
     }
+    ExpressionEdge eAttributes = new ExpressionEdge(vAttributes, null);
+    exprGraph.addEdge(eAttributes);
+    PrimitivePortVertex vPort = new PrimitivePortVertex(exprGraph,
+        eSignalType, eAttributes);
+    exprGraph.addVertex(vPort);
+    return vPort;
   }
 
   @Override
-  public Expression visitTupleTypeValue(TupleTypeValueContext context) {
-    // visit all children
+  public ExpressionVertex visitTupleTypeValue(TupleTypeValueContext context) {
     List<TupleTypeValueEntryContext> entries = context.tupleTypeValueEntry();
-    Map<String, Expression> typeExprs = new HashMap<>();
-    Map<String, Expression> defaultValues = new HashMap<>();
+    Map<String, ExpressionEdge> typeValueEdges = new HashMap<>();
+    Map<String, ExpressionEdge> defaultValueEdges = new HashMap<>();
     for (TupleTypeValueEntryContext entryCtx : entries) {
       // each child has a typevalue, and may have
       // an identifier (named field)
       // and an expression (default value)
-      Expression typevalue = entryCtx.typevalue().accept(this);
       String identifier;
       Integer nextAnonymousID = 0;
```

Comment:
Shouldn't this be declared outside of the loop? It's always going to generate identifier="0" .
Also,  `nextAnonymousID += 1` currently doesn't do anything..